File: python/quick_sort.py
from random import randint
import time

start = time.time()
def quicksort(array: list[int]) -> list[int]:
    if len(array) < 2: return array

    low, same, high = [], [], []

    pivot = array[randint(0, len(array) - 1)]
    for item in array:
        if item < pivot:
            low.append(item)

        elif item == pivot:
            same.append(item)

        elif item > pivot:
            high.append(item)


    return quicksort(low) + same + quicksort(high)

def gen_random_arr(length: int, min: int, max: int) -> list[int]:
    res = []
    for _ in range(length):
        res.append(randint(min, max+1))

    return res


print(f"Total time taken: {time.time() - start}")

File: src/shah/jeevan/QuickSort.java
package shah.jeevan;

import java.util.ArrayList;
import java.util.Random;

public class QuickSort {
    public static ArrayList<Integer> quickSort(ArrayList<Integer> arr) {
        Random rand = new Random();

        if (arr.size() < 2) { return arr; }

        ArrayList<Integer> low = new ArrayList<>();
        ArrayList<Integer> same = new ArrayList<>();
        ArrayList<Integer> high = new ArrayList<>();

        int pivot = rand.nextInt(arr.size() - 1);

        for (int i : arr) {
            if (i < pivot) {
                low.add(i);
            } else if (i == pivot) {
                same.add(i);
            } else if (i > pivot) {
                high.add(i);
            }
        }
        ArrayList<Integer> res = new ArrayList<>(quickSort(low));
        res.addAll(same);
        res.addAll(quickSort(high));

        return res;
    }
}
